crates/ruma-api-macros/src/request.rs

```diff
@@ -164,8 +164,27 @@ impl Request {
         self.fields.iter().filter(|f| matches!(f, RequestField::Header(..)))
     }
 
-    fn path_fields(&self) -> impl Iterator<Item = &Field> {
-        self.fields.iter().filter_map(RequestField::as_path_field)
+    fn path_fields_ordered(&self) -> impl Iterator<Item = &Field> {
+        let map: BTreeMap<String, &Field> = self
+            .fields
+            .iter()
+            .filter_map(RequestField::as_path_field)
+            .map(|f| (f.ident.as_ref().unwrap().to_string(), f))
+            .collect();
+
+        self.stable_path
+            .as_ref()
+            .or(self.r0_path.as_ref())
+            .or(self.unstable_path.as_ref())
+            .expect("one of the paths to be defined")
+            .value()
+            .split('/')
+            .filter_map(|s| {
+                s.strip_prefix(':')
+                    .map(|s| *map.get(s).expect("path args have already been checked"))
+            })
+            .collect::<Vec<_>>()
+            .into_iter()
```

Comment:
I think you shouldn't need to collect to a temporary `Vec` here. Does this work?

```suggestion
            .filter_map(move |s| {
                s.strip_prefix(':')
                    .map(|s| *map.get(s).expect("path args have already been checked"))
            })
```

Reply:
Nope, that's why we're collecting in the first place, because I encountered this problem myself while writing this.

It shouldn't affect performance too much, though i'll try to find something better.